File: benchmarks/php/mandelbrot.php
#!/usr/bin/env php
<?php
// Mandelbrot set calculation benchmark

// Size constants
define('WIDTH', 800);
define('HEIGHT', 800);
define('MAX_ITERATIONS', 100);

// Calculate the Mandelbrot set
function calculate_mandelbrot($max_iterations) {
    $result = array();
    
    for ($y = 0; $y < HEIGHT; $y++) {
        $result[$y] = array();
        for ($x = 0; $x < WIDTH; $x++) {
            $cx = ($x - WIDTH / 2.0) * 4.0 / WIDTH;
            $cy = ($y - HEIGHT / 2.0) * 4.0 / HEIGHT;
            
            $zx = 0.0;
            $zy = 0.0;
            $iteration = 0;
            
            while ($zx*$zx + $zy*$zy <= 4.0 && $iteration < $max_iterations) {
                $temp = $zx*$zx - $zy*$zy + $cx;
                $zy = 2.0 * $zx * $zy + $cy;
                $zx = $temp;
                $iteration++;
            }
            
            $result[$y][$x] = $iteration;
        }
    }
    
    return $result;
}

// Count non-black pixels
function count_non_black($result, $max_iterations) {
    $count = 0;
    
    for ($y = 0; $y < HEIGHT; $y++) {
        for ($x = 0; $x < WIDTH; $x++) {
            if ($result[$y][$x] < $max_iterations) {
                $count++;
            }
        }
    }
    
    return $count;
}

// Main function
function main() {
    $start_time = microtime(true);
    
    $result = calculate_mandelbrot(MAX_ITERATIONS);
    $count = count_non_black($result, MAX_ITERATIONS);
    
    echo "Mandelbrot set calculation finished.\n";
    echo "Image size: " . WIDTH . " x " . HEIGHT . "\n";
    echo "Maximum iterations: " . MAX_ITERATIONS . "\n";
    echo "Non-black pixels: " . $count . "\n";
    
    $elapsed = (microtime(true) - $start_time) * 1000;
    echo "Time taken: $elapsed ms\n";
    
    // Get memory stats
    $memoryUsed = memory_get_peak_usage(true) / 1024;
    echo "Memory used: $memoryUsed KB\n";
}

main();

File: benchmarks/perl/mandelbrot.pl
#!/usr/bin/env perl
# Mandelbrot set calculation benchmark

use strict;
use warnings;
use Time::HiRes qw(gettimeofday tv_interval);

# Size constants
use constant WIDTH => 800;
use constant HEIGHT => 800;
use constant MAX_ITERATIONS => 100;

# Calculate the Mandelbrot set
sub calculate_mandelbrot {
    my ($max_iterations) = @_;
    my @result;
    
    for my $y (0..HEIGHT-1) {
        for my $x (0..WIDTH-1) {
            my $cx = ($x - WIDTH / 2.0) * 4.0 / WIDTH;
            my $cy = ($y - HEIGHT / 2.0) * 4.0 / HEIGHT;
            
            my $zx = 0.0;
            my $zy = 0.0;
            my $iteration = 0;
            
            while ($zx*$zx + $zy*$zy <= 4.0 && $iteration < $max_iterations) {
                my $temp = $zx*$zx - $zy*$zy + $cx;
                $zy = 2.0 * $zx * $zy + $cy;
                $zx = $temp;
                $iteration++;
            }
            
            $result[$y][$x] = $iteration;
        }
    }
    
    return \@result;
}

# Count non-black pixels
sub count_non_black {
    my ($result, $max_iterations) = @_;
    my $count = 0;
    
    for my $y (0..HEIGHT-1) {
        for my $x (0..WIDTH-1) {
            if ($result->[$y][$x] < $max_iterations) {
                $count++;
            }
        }
    }
    
    return $count;
}

# Main
my $start_time = [gettimeofday];

my $result = calculate_mandelbrot(MAX_ITERATIONS);
my $count = count_non_black($result, MAX_ITERATIONS);

print "Mandelbrot set calculation finished.\n";
printf "Image size: %d x %d\n", WIDTH, HEIGHT;
printf "Maximum iterations: %d\n", MAX_ITERATIONS;
printf "Non-black pixels: %d\n", $count;

my $elapsed = tv_interval($start_time) * 1000;
printf "Time taken: %.2f ms\n", $elapsed;

# Perl doesn't have a standard way to get memory usage
print "Memory monitoring not available for Perl implementation\n";